pkg/container/runc/img.go

```diff
@@ -0,0 +1,266 @@
+	"github.com/containerd/containerd/namespaces"
+	"github.com/docker/distribution/reference"
+	controlapi "github.com/moby/buildkit/api/services/control"
+	bkclient "github.com/moby/buildkit/client"
+	"github.com/moby/buildkit/cmd/buildctl/build"
+	"github.com/moby/buildkit/identity"
+	"github.com/moby/buildkit/session"
+	"github.com/moby/buildkit/util/progress/progressui"
+	"golang.org/x/sync/errgroup"
+
+	"github.com/dominodatalab/forge/pkg/archive"
+	"github.com/dominodatalab/forge/pkg/container/config"
+	imgclient "github.com/dominodatalab/forge/pkg/img/client"
+	"github.com/dominodatalab/forge/pkg/img/types"
+)
+
+type Builder struct {
+	client *imgclient.Client
+}
+
+func NewImgBuilder() (*Builder, error) {
+	c, err := imgclient.New(getStateDirectory(), types.AutoBackend)
+	if err != nil {
+		return nil, err
+	}
+
+	return &Builder{client: c}, nil
+}
+
+func (b *Builder) Build(ctx context.Context, opts config.BuildOptions) (string, error) {
+	image, err := b.build(ctx, opts)
+	if err != nil {
+		return "", err
+	}
+
+	if opts.SizeLimit != 0 {
+		if err := b.validateImageSize(ctx, image, opts.SizeLimit); err != nil {
+			return "", err
+		}
+	}
+
+	if err := b.push(ctx, image); err != nil {
+		return "", err
+	}
+	return image, nil
+}
+
+func (b *Builder) build(ctx context.Context, opts config.BuildOptions) (string, error) {
+	// download and extract remote OCI context
+	extract, err := archive.FetchAndExtract(opts.Context)
+	if err != nil {
+		return "", err
+	}
+	defer os.RemoveAll(extract.RootDir)
+
+	// assume Dockerfile lives inside context root
+	localDirs := map[string]string{
+		"context":    extract.ContentsDir,
+		"dockerfile": extract.ContentsDir,
+	}
+
+	// create a new buildkit session
+	sess, sessDialer, err := b.client.Session(ctx, localDirs)
+	if err != nil {
+		return "", err
+	}
+
+	// prepare build parameters
+	solveReq, err := solveRequestWithContext(sess.ID(), opts)
+	if err != nil {
+		sess.Close()
+		return "", err
+	}
+
+	// add build metadata to context
+	ctx = session.NewContext(ctx, sess.ID())
+	ctx = namespaces.WithNamespace(ctx, "buildkit")
+	eg, ctx := errgroup.WithContext(ctx)
+
+	// launch build
+	ch := make(chan *controlapi.StatusResponse)
+	eg.Go(func() error {
+		return sess.Run(ctx, sessDialer)
+	})
+	eg.Go(func() error {
+		defer sess.Close()
+		return b.client.Solve(ctx, solveReq, ch)
+	})
+	eg.Go(func() error {
+		return showProgress(ch, false)
+	})
+	if err := eg.Wait(); err != nil {
+		return "", err
+	}
+
+	// return final image url
+	return solveReq.ExporterAttrs["name"], nil
+}
+
+func (b *Builder) validateImageSize(ctx context.Context, name string, limit uint64) error {
+	id := identity.NewID()
+	ctx = session.NewContext(ctx, id)
+	ctx = namespaces.WithNamespace(ctx, "buildkit")
+
+	images, err := b.client.ListImages(ctx, fmt.Sprintf("name==%s", name))
+	if err != nil {
+		return err
+	}
+	if len(images) != 1 {
+		return fmt.Errorf("could not find exact image %q in list: %v", name, images)
+	}
```

Comment:
Seems a little strange to `List` and then only look for one, it looks like the ImageStore supports a `Get`?